codec/src/main/java/io/netty/handler/codec/protobuf/ProtobufVarint32FrameDecoder.java

```diff
@@ -15,63 +15,18 @@
  */
 package io.netty.handler.codec.protobuf;
 
-import io.netty.buffer.ByteBuf;
-import io.netty.channel.ChannelHandlerContext;
-import io.netty.handler.codec.ByteToMessageDecoder;
-import io.netty.handler.codec.CorruptedFrameException;
+import com.google.protobuf.CodedInputStream;
 
-import java.util.List;
+import java.io.IOException;
 
-import com.google.protobuf.CodedInputStream;
+import io.netty.handler.codec.protobuf.util.AbstractProtobufVarint32FrameDecoder;
 
 /**
- * A decoder that splits the received {@link ByteBuf}s dynamically by the
- * value of the Google Protocol Buffers
- * <a href="http://code.google.com/apis/protocolbuffers/docs/encoding.html#varints">Base
- * 128 Varints</a> integer length field in the message.  For example:
- * <pre>
- * BEFORE DECODE (302 bytes)       AFTER DECODE (300 bytes)
- * +--------+---------------+      +---------------+
- * | Length | Protobuf Data |----->| Protobuf Data |
- * | 0xAC02 |  (300 bytes)  |      |  (300 bytes)  |
- * +--------+---------------+      +---------------+
- * </pre>
- *
- * @see CodedInputStream
+ * {@link AbstractProtobufVarint32FrameDecoder}
  */
-public class ProtobufVarint32FrameDecoder extends ByteToMessageDecoder {
-
-    // TODO maxFrameLength + safe skip + fail-fast option
-    //      (just like LengthFieldBasedFrameDecoder)
-
+public class ProtobufVarint32FrameDecoder extends AbstractProtobufVarint32FrameDecoder {
     @Override
-    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
-        in.markReaderIndex();
-        final byte[] buf = new byte[5];
-        for (int i = 0; i < buf.length; i ++) {
-            if (!in.isReadable()) {
-                in.resetReaderIndex();
-                return;
-            }
-
-            buf[i] = in.readByte();
-            if (buf[i] >= 0) {
-                int length = CodedInputStream.newInstance(buf, 0, i + 1).readRawVarint32();
-                if (length < 0) {
-                    throw new CorruptedFrameException("negative length: " + length);
-                }
-
-                if (in.readableBytes() < length) {
-                    in.resetReaderIndex();
-                    return;
-                } else {
-                    out.add(in.readBytes(length));
-                    return;
-                }
-            }
-        }
-
-        // Couldn't find the byte whose MSB is off.
-        throw new CorruptedFrameException("length wider than 32-bit");
+    protected int readRawVarint32(byte[] buf, int off, int len) throws IOException {
+        return CodedInputStream.newInstance(buf, off, len).readRawVarint32();
```

Comment:
yeah i considered doing it but though it'd be considered bad practice.  If they for some reason ever change their encoding this would break. I agree that it feels really ugly to do it this way but given the way protobufs code i don't see any other way, excluding copying code from the library.

Does the neater design outweigh the potential break? Chances are they will redesign the library before they change the encoding but I'm still reluctant to copy code. WDYT?